lua/lspconfig/ltex.lua

```diff
@@ -0,0 +1,186 @@
+    for _, lsp in ipairs(buf_clients) do
+        if lsp.name == "ltex" then
+            client = lsp
+        end
+    end
+
+    if client then
+        if configtype == 'dictionary' then
+            -- if client.config.settings.ltex.dictionary then
+                client.config.settings.ltex.dictionary = {
+                    [lang] = readFiles(client.config.dictionary_files[lang])
+                };
+                return client.notify('workspace/didChangeConfiguration', client.config.settings)
+            -- else
+                -- return vim.notify("Error when reading dictionary config, check it")
+            -- end
+        elseif configtype == 'disable' then
+            if client.config.settings.ltex.disabledRules then
+                client.config.settings.ltex.disabledRules = {
+                    [lang] = readFiles(client.config.disabledrules_files[lang])
+                };
+                return client.notify('workspace/didChangeConfiguration', client.config.settings)
+            else
+                return vim.notify("Error when reading disabledRules config, check it")
+            end
+
+        elseif configtype == 'falsePositive' then
+            if client.config.settings.ltex.hiddenFalsePositives then
+                client.config.settings.ltex.hiddenFalsePositives = {
+                    [lang] = readFiles(client.config.falsepositive_files[lang])
+                };
+                return client.notify('workspace/didChangeConfiguration', client.config.settings)
+            else
+                return vim.notify("Error when reading hiddenFalsePositives config, check it")
+            end
+        end
+    else
+        return nil
+    end
+end
+
+local function addToFile(filetype, lang, file, value)
+    file = io.open(file[#file-0], "a+") -- add only to last file defined.
+    if file then
+        file:write(value .. "\n")
+        file:close()
+    else
+        return print("Failed insert %q", value)
+    end
+    if filetype == 'dictionary' then
+        return updateConfig(lang, "dictionary")
+    elseif filetype == 'disable' then
+        return updateConfig(lang, "disable")
+    elseif filetype == 'falsePositive' then
+        return updateConfig(lang, "falsePositive")
+    end
+end
+
+local function addTo(filetype, lang, file, value)
+    local dict = readFiles(file)
+    for _, v in ipairs(dict) do
+        if v == value then
+            return nil
+        end
+    end
+    return addToFile(filetype, lang, file, value)
+end
+
+configs.ltex = {
+    default_config = {
+        cmd = {"ltex-ls"};
+        filetypes = {'tex', 'bib', 'md'};
+        root_dir = function(filename)
+            return util.path.dirname(filename)
+        end;
+        dictionary_files = { ["en"] = {vim.fn.getcwd() .. "dictionary.ltex"} };
+        disabledrules_files = { ["en"] = {vim.fn.getcwd() .. "disable.ltex"} };
+        falsepositive_files = { ["en"] = {vim.fn.getcwd() .. "false.ltex"}};
+        settings = {
```

Comment:
I'm pretty sure everything under settings is either a default, or too opinionated to be worth using.

Reply:
> With the size of this my module, my recommendation is to create ltex-ls.nvim and add a smaller minimal config in lspconfig to start. The following "works". We can gradually add back in the functionality while making any necessary changes in neovim core to support more "off-spec" language servers like this one:
> 
> ```lua
> local configs = require'lspconfig/configs'
> local util = require 'lspconfig/util'
> 
> local server_name = "ltex-ls"
> 
> configs[server_name] = {
>   default_config = {
>     cmd = {"ltex-ls"};
>     filetypes = {'tex', 'bib', 'markdown'};
>     root_dir = function(filename)
>       return util.path.dirname(filename)
>     end;
>     settings = {
>       ltex = {
>         checkFrequency="edit",
>       };
>     };
>   };
>   docs = {
>     package_json = 'https://raw.githubusercontent.com/valentjn/vscode-ltex/develop/package.json',
>     description = [[
> https://github.com/valentjn/ltex-ls
> 
> LTeX Language Server: LSP language server for LanguageTool 🔍✔️ with support for LaTeX 🎓, Markdown 📝, and others
> 
> To install, download the latest [release](https://github.com/valentjn/ltex-ls/releases) and ensure `ltex-ls` is on your path.
> 
> ]];
>   };
> };
> ```
> 
> edit: Actually we need to send at least one setting for the server to trigger... seems like a bug.

What exactly didn't work? At least in my test, except to the "add to ..." and pickyrules which is disabled by default.

Reply:
The server doesn't provide any diagnostics if not at least one `settings` item is sent, I'm advocating for the first round of inclusion of lspconfig, using the simpler config you quoted while we clean up this PR.